### ISSUE TITLE
Using config-rs for layered configuration

### ISSUE THREAD

**Issue by bIgBV** [rust-net-web/tide#123]
Now that #109 has landed, I am looking to extend it so that there are other possible ways of setting up the execution environment for a Tide application. When it comes to application deployments, the [12-factor app](https://12factor.net/) idea is standard.

The [config-rs](https://docs.rs/config/0.9.2/config/) crate provides good support for layering configuration from different sources. It also seems to be the most downloaded crate in the configuration category according [crates.rs](https://crates.rs/config) (To be fair, that metric is pretty short sighted).

The issue is that it makes the API for accessing application configuration becomes a little verbose. I feel that the flexibility afforded is worth is.

What are your thoughts on this?

**Comment by yoshuawuyts** [comment]
> The twelve-factor app stores config in environment variables (often shortened to env vars or env). Env vars are easy to change between deploys without changing any code; unlike config files, there is little chance of them being checked into the code repo accidentally (...)

@bIgBV doesn't the 12 factor app standard disencourage the use of config files? From the docs `config-rs` seems to take in configuration from a lot of places, only one of which is the environment.

In my Rust applications I've been using `structopt` to define both flags and environment variables. For example: https://github.com/rust-clique/clap-port-flag/blob/master/src/lib.rs#L38.

I think it might be useful to suggest methods of configuration with Tide, but probably not to build into the framework directly.

**Comment by yoshuawuyts** [comment]
> It just has to provide the mechanisms for arbitrary config data.

@PeteJodo The way I think about this relationship is:
- structopt gets data from the CLI, either from flags or env vars
- in the main fn this data gets passed into Tide's app + middleware constructors. This happens only once, at boot
- every request now is passed through middleware, and each `Store` (formerly `Config`) is populated with the corresponding data

---

> Tide really only relies on host and port

Structopt supports reusing flags! For address & port: https://github.com/rust-clique/clap-port-flag

**Comment by bIgBV** [comment]
@yoshuawuyts you're right, it does discourage config files, I figured it wouldn't be too difficult to add support for that on top of CLI flags and env vars. Rocket uses a config file and I thought it would be useful to have a similar mechanism.

On the other hand, I did not know about structopt and I think it would be a really handy way of managing configuration. 

> * structopt gets data from the CLI, either from flags or env vars
> * in the main fn this data gets passed into Tide's app + middleware constructors. This happens only once, at boot
> * every request now is passed through middleware, and each Store (formerly Config) is populated with the corresponding 

I really like this approach, and I'll open a PR exploring this. I don't mind adding a dependency since the end user can overwrite the `Config` struct in the `Store` if they want to modify something at runtime anyway.

**Comment by prasannavl** [comment]
Hmm. I looked at #129, as well. I think this entire area of exploration is increasing complexity and creating a domain specific API while in practicality this probably is not at all needed. Once  https://github.com/rustasync/tide/pull/156 lands, as long as `Configuration` is done in a generic way where any type can go it, and be taken out, I feel that's the extent that `tide` should go, beyond some tide specific config that becomes a part of the `ConfigurationBuilder` (or perhaps a specific conversion mechanism in an ergonomic way if that becomes tedious). 

Right now, one of the reasons I beginning to like the direction of tide (esp. with #156), is that it focuses on simplicity and not re-creating concepts or ideas in the name of features. I think configuration is better solved outside of tide, with tide just taking in what's provided from the outside.  

I feel bring in `config-rs` or any specific way of configuration for that matter should be avoided. Today `config-rs` might be a popular way, tomorrow it may be something else, but tide locking into one of these patterns is both unnecessary and prohibitive.

**Comment by mehcode** [comment]
Just want to pop in and say that I would be disappointed if a library like this were to process configuration files or environment variables on its own. I very much agree with @prasannavl here.

**Comment by yoshuawuyts** [comment]
Unrelated (but fun!) update on the structopt side: we're currently experimenting with what shape it could take if we could make command line parsing _"more first class"_ in the language.

https://github.com/rust-cli/paw is where we're currently experimenting, and if it turns out alright, I think it'll nicely compliment Tide (as a separate package :sparkles:).

## Vaporware Example
```rust
use std::net::TcpListener;

#[paw_clap]
struct Args {
    /// The port to listen on.
    port: usize,
};

async fn echo_string(mut cx: Context<()>) -> String {
    let msg = await!(cx.body_string()).unwrap();
    println!("String: {}", msg);
    msg
}

#[paw::main]
fn main(args: Args) -> Result<(), failure::Error> {
    let mut app = App::new(());
    app.at("/echo/string").post(echo_string);
    app.serve(("127.0.0.1", args.port))?;
}
```

**Comment by aturon** [comment]
Now that #156 has landed, we should take a fresh look at configuration.

**Comment by PeteJodo** [comment]
this issue looks outdated. Everything that's needed for configuration, namely host and port, is up to the user. However, that seemed to have been the case after #156 and @aturon's comment anyway where he suggested taking a fresh look at configuration so maybe I'm missing something?

**Comment by yoshuawuyts** [comment]
We may need to allow for more elaborate configuration still. One way would be to make it part of `listen`, where a string can convert itself to configuration, but we can also take an initialized configuration object. More or less like the [pattern API](https://doc.rust-lang.org/core/str/pattern/trait.Pattern.html#implementors) works in std.
